### PR TITLE
Ner crf v2 request2

### PR DESCRIPTION
**Pull request by jayaraopratik**

### CODE REVIEW

File: chatbot_ner/config.py [hunk 1]
```diff
@@ -85,6 +95,12 @@
 ELASTICSEARCH_CRF_DATA_INDEX_NAME = os.environ.get('ELASTICSEARCH_CRF_DATA_INDEX_NAME')
 ELASTICSEARCH_CRF_DATA_DOC_TYPE = os.environ.get('ELASTICSEARCH_CRF_DATA_DOC_TYPE')
 
+
+# Crf Model Specific with additional AWS storage (optional)
+AWS_MODEL_BUCKET = os.environ.get('AWS_MODEL_BUCKET')
```
Comment (amansrivastava17) [inline chatbot_ner/config.py]: Rename this CRF_MODEL_S3_BUCKET_NAME

File: chatbot_ner/config.py [hunk 2]
```diff
@@ -85,6 +95,12 @@
...
+AWS_MODEL_REGION = os.environ.get('AWS_MODEL_REGION')
```
Comment (amansrivastava17) [inline chatbot_ner/config.py]: Rename this CRF_MODEL_S3_BUCKET_REGION

File: chatbot_ner/config.py [hunk 3]
```diff
@@ -85,6 +95,12 @@
...
+
+REMOTE_EMBEDDINGS_URL = os.environ.get('REMOTE_EMBEDDINGS_URL')
```
Comment (amansrivastava17) [inline chatbot_ner/config.py]: Rename this to WORD_EMBEDDING_REMOTE_URL

File: chatbot_ner/settings.py [hunk 4]
```diff
@@ -10,8 +10,13 @@
 
 # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
 import os
+import dotenv
 
 BASE_DIR = os.path.dirname(os.path.dirname(__file__))
+CONFIG_PATH = os.path.join(BASE_DIR, 'config')
+
```
Comment (amansrivastava17) [inline chatbot_ner/settings.py]: Remove this part, not required

File: chatbot_ner/urls.py [hunk 5]
```diff
@@ -32,4 +32,10 @@
     url(r'^entities/get_crf_training_data', 'external_api.api.get_crf_training_data'),
     url(r'^entities/update_crf_training_data', 'external_api.api.update_crf_training_data'),
 
+    #  Train Crf Model
+    url(r'^entities/train_crf_model', 'external_api.api.train_crf_model'),
+
+    #  Transfer Crf Model
+    url(r'^entities/transfer_crf_model', 'external_api.api.transfer_crf_model'),
```
Comment (amansrivastava17) [inline chatbot_ner/urls.py]: remove

File: external_api/api.py [hunk 6]
```diff
@@ -10,8 +10,10 @@
     FetchIndexForAliasException, DeleteIndexFromAliasException
 from chatbot_ner.config import ner_logger
 from external_api.constants import ENTITY_DATA, ENTITY_NAME, LANGUAGE_SCRIPT, ENTITY_LIST, \
-    EXTERNAL_API_DATA, SENTENCE_LIST
+    EXTERNAL_API_DATA, SENTENCE_LIST, CLOUD_STORAGE, ES_CONFIG, CLOUD_EMBEDDINGS, LIVE_CRF_MODEL_PATH
```
Comment (amansrivastava17) [inline external_api/api.py]: rename this `CLOUD_STORAGE` and `CLOUD_EMBEDDINGS`

File: models/crf_v2/crf_detect_entity.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+class CrfDetection(object):
+    """
+    This method is used to detect a text entity using the Crf model.
+    """
+    def __init__(self, entity_name, cloud_storage=False, cloud_embeddings=False, live_crf_model_path=''):
+        """
+        This method is used to detect text entities using the Crf model
+        Args:
+            entity_name (str): Name of the entity for which the entity has to be detected
+            cloud_storage (bool): To indicate if cloud storage settings is required.
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_detect_entity.py]: doc

File: models/crf_v2/crf_preprocess_data.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+from models.crf_v2.constants import TEXT_LIST, WORD_EMBEDDINGS, WORD_VEC_FEATURE, B_LABEL,\
+    B_TAG, I_LABEL, I_TAG, POS_TAGS, LABELS, O_LABEL, BOS, EOS
+
+
+class CrfPreprocessData(object):
+    """
+    This class is used to pre_process_data for the Crf model.
+    """
+    @staticmethod
+    def pre_process_text_(text, entities):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+                          LABELS: []}
+
+        for text, entities in zip(text_list, entity_list):
+            tokenzied_text, labels = CrfPreprocessData.pre_process_text_(text, entities)
+            processed_list[TEXT_LIST].append(labels)
+            processed_list[LABELS].append(tokenzied_text)
+        return processed_list
+
+    @staticmethod
+    def pos_tag(docs):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+        Returns:
+            (list): List consisting of the features used to train the CRF model.
+        """
+        features = []
+        for i in range(len(doc[TEXT_LIST])):
+            features.append([CrfPreprocessData.word_to_features(doc, i, j) for j in range(len(doc[TEXT_LIST][i]))])
+        return features
+
+    @staticmethod
+    def get_processed_x_y(text_list, entity_list=[[]], cloud_embeddings=False):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+            word_embedding = doc[WORD_EMBEDDINGS][i][j + 1]
+            features.extend(CrfPreprocessData.convert_wordvec_features('+1', word_embedding))
+        else:
+            # Indicate that it is the 'end of a document'
+            features.append(EOS)
+
+        return features
+
+    @staticmethod
+    def extract_features(doc):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+             convert_wordvec_features(prefix, word_vec)
+             >> ['-1word_vec0=0.23', '-1word_vec1=0.45', '-1word_vec2=0.11']
+        """
+        features = []
+        for i, each in enumerate(word_vec):
+            features.append(prefix + WORD_VEC_FEATURE + str(i) + '=' + str(each))
+        return features
+
+    @staticmethod
+    def word_to_features(doc, i, j):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+        """
+        docs[POS_TAGS] = []
+        pos_tagger = POS()
+        for text in docs[TEXT_LIST]:
+            docs[POS_TAGS].append([tag[1] for tag in pos_tagger.tagger.tag(text)])
+
+        return docs
+
+    @staticmethod
+    def convert_wordvec_features(prefix, word_vec):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+        features = CrfPreprocessData.extract_features(processed_text_pos_tag)
+        ner_logger.debug('CrfPreprocessData.extract_features Completed')
+
+        ner_logger.debug('CrfPreprocessData.get_labels Started')
+        labels = processed_text_pos_tag['labels']
+        ner_logger.debug('CrfPreprocessData.get_labels Completed')
+        return features, labels
+
+    @staticmethod
+    def remote_word_embeddings(text_dict):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: done

File: models/crf_v2/crf_train.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            self.write_model_to_s3()
+            return self.model_dir
+        else:
+            local_path = MODELS_PATH + self.entity_name
+            trainer.train(local_path)
+            ner_logger.debug('Training for entity %s completed' % self.entity_name)
+            ner_logger.debug('Model locally saved at %s' % self.entity_name)
+            return local_path
+
+    def train_model(self, text_list, entity_list, c1=0, c2=0, max_iterations=1000):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_train.py]: The train_crf_model is used to define the hyperparameters of the model.
While train_model is for pre processing  and any additional functions and run the model defined in train_crf_model

File: models/crf_v2/crf_train.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        if not entity_list:
+            raise ESCrfTrainingEntityListNotFoundException()
+
+        ner_logger.debug('Fetch of data from ES for ENTITY: %s completed' % self.entity_name)
+        ner_logger.debug('Length of text_list %s' % str(len(text_list)))
+
+        model_path = self.train_model(entity_list=entity_list, text_list=text_list)
+        return model_path
+
+    def write_model_to_s3(self):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_train.py]: done

File: models/crf_v2/crf_train.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+                                  bucket_region=CRF_MODEL_S3_BUCKET_REGION,
+                                  address=self.model_dir,
+                                  disk_filepath=self.model_dir)
+        if result:
+            ner_logger.debug('Model : %s written to s3' % self.model_dir)
+        else:
+            ner_logger.debug('Failure in saving Model to s3 %s' % self.model_dir)
+            raise AwsCrfModelWriteException()
+
+    def generate_model_path(self):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_train.py]: done

File: models/crf_v2/word_embeddings.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/word_embeddings.py]: done

File: models/crf_v2/word_embeddings.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    @staticmethod
+    def load_word_vectors_local():
+        """
+        Thus function is used to load the word_list and word_vectors from the specified paths.
+        Returns:
+        vocab (list): word_list present at the specified path.
+        word_vectors (numpy.array): word_vectors present at the specified path.
+        """
+        file_handler = open(EMBEDDINGS_PATH_VOCAB, 'rb')
+        vocab = pickle.load(file_handler)
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/word_embeddings.py]: done

File: models/crf_v2/word_embeddings.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    def load_word_vectors_remote(text_list):
+        """
+        This method is used to load word_vectors from a remote location
+        Args:
+            text_list (list): List of tokenized texts
+            text_list = [['my', 'name', 'is', 'chatbot'], ['how', 'are', 'you']]
+        Returns:
+            word_vectors (np.ndarray): Numpy array of vectors for the provided text_list
+        """
+        json_dict = {TEXT_LIST: [text_list]}
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/word_embeddings.py]: done

File: models/crf_v2/crf_preprocess_data.py [hunk 21]
```diff
@@ -0,0 +1,368 @@
...
+            processed_data (list): Returns a list of tuples where each tuple is of the form (token, label)
+
+        Example:
+            text_list ='book a flight from Mumbai to Delhi'
+            entity_list = ['Mumbai', 'Delhi']
+            pre_process_text_(text, entity_list)
+            >> ['book', 'a', 'flight', 'from', 'Mumbai', 'to', 'Delhi'],
+                ['O', 'O', 'O', 'O', 'B', 'O', 'B']
+        """
+        def iob_prefixes(entity_value, word_tokenize):
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/crf_preprocess_data.py]: This is done for readability purposes. And easier to debug

File: external_api/constants.py [hunk 22]
```diff
@@ -4,3 +4,7 @@
 LANGUAGE_SCRIPT = 'language_script'
 ENTITY_LIST = 'entity_list'
 SENTENCE_LIST = 'sentence_list'
+CLOUD_STORAGE = 'cloud_storage'
+ES_CONFIG = 'es_config'
+CLOUD_EMBEDDINGS = 'cloud_embeddings'
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply external_api/constants.py]: done

File: ner_v1/api.py [hunk 23]
```diff
@@ -40,6 +40,9 @@ def get_parameters_dictionary(request):
                        PARAMETER_MIN_TOKEN_LEN_FUZZINESS: request.GET.get('min_token_len_fuzziness'),
                        PARAMETER_MIN_DIGITS: request.GET.get('min_number_digits'),
                        PARAMETER_MAX_DIGITS: request.GET.get('max_number_digits'),
+                       PARAMETER_CLOUD_EMBEDDINGS: request.GET.get('cloud_embeddings', 'False'),
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply ner_v1/api.py]: done

File: ner_v1/constant.py [hunk 24]
```diff
@@ -73,3 +73,10 @@
 PARAMETER_FUZZINESS = 'fuzziness'
 PARAMETER_MIN_TOKEN_LEN_FUZZINESS = 'min_token_len_fuzziness'
 DATASTORE_VERIFIED = 'datastore_verified'
+CRF_MODEL_VERIFIED = 'crf_model_verified'
+
+#  **********************constants used for text detection************************************
+
+PARAMETER_CLOUD_STORAGE = 'cloud_storage'
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply ner_v1/constant.py]: done

File: external_api/constants.py [hunk 25]
```diff
@@ -4,3 +4,7 @@
 LANGUAGE_SCRIPT = 'language_script'
 ENTITY_LIST = 'entity_list'
 SENTENCE_LIST = 'sentence_list'
+CLOUD_STORAGE = 'cloud_storage'
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply external_api/constants.py]: done

File: external_api/api.py (hunk 6 above)
Comment (jayaraopratik) [reply external_api/api.py]: done

File: models/crf_v2/constants.py [hunk 26]
```diff
@@ -0,0 +1,15 @@
+ENTITY_LIST = 'entity_list'
+SENTENCE_LIST = 'sentence_list'
+TEXT_LIST = 'text_list'
```
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply models/crf_v2/constants.py]: done

File: external_api/constants.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply external_api/constants.py]: done

File: external_api/constants.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply external_api/constants.py]: done

File: external_api/api.py (hunk 6 above)
Comment (jayaraopratik) [reply external_api/api.py]: done

File: ner_v1/api.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply ner_v1/api.py]: done

File: ner_v1/constant.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (jayaraopratik) [reply ner_v1/constant.py]: done

File: chatbot_ner/config.py [hunk 27]
```diff
@@ -71,6 +72,11 @@
 ES_BULK_MSG_SIZE = os.environ.get('ES_BULK_MSG_SIZE', '10000')
 ES_SEARCH_SIZE = os.environ.get('ES_SEARCH_SIZE', '10000')
 
+# Crf Model Specific (Mandatory to use CRF Model)
+MODELS_PATH = os.environ.get('MODELS_PATH')
```
Comment (amansrivastava17) [inline chatbot_ner/config.py]: add crf_prefix

File: external_api/constants.py [hunk 28]
```diff
@@ -4,3 +4,7 @@
 LANGUAGE_SCRIPT = 'language_script'
 ENTITY_LIST = 'entity_list'
 SENTENCE_LIST = 'sentence_list'
+READ_MODEL_FROM_S3 = 'read_model_from_s3'
```
Comment (amansrivastava17) [inline external_api/constants.py]: add crf_ prefix

**Pull request merged**